internal/codegen/blocks.go

```diff
@@ -105,51 +126,177 @@ func parseBlocks(filePath string) (map[string]string, error) {
 			switch cmd {
 			case "Block":
 				blockName := matches[3]
-				if curBlockName != "" {
+				if curBlock != nil {
 					return nil, fmt.Errorf("invalid Block when already inside of a block, at %s:%d", filePath, i+1)
 				}
-				curBlockName = blockName
+				curBlock = &blockInfo{
+					Name:      blockName,
+					StartLine: i,
+				}
+				blocks[blockName] = curBlock
 			case endStatement:
 				blockName := matches[3]
 
-				if curBlockName == "" {
+				if curBlock == nil {
 					return nil, fmt.Errorf("invalid EndBlock when not inside of a block, at %s:%d", filePath, i+1)
 				}
 
-				if blockName != curBlockName {
+				if blockName != curBlock.Name {
 					return nil, fmt.Errorf(
 						"invalid EndBlock, found EndBlock with name %q while inside of block with name %q, at %s:%d",
-						blockName, curBlockName, filePath, i+1,
+						blockName, curBlock.Name, filePath, i+1,
 					)
 				}
 
-				curBlockName = ""
+				curBlock.EndLine = i
+				curBlock = nil
 			default:
 				isCommand = false
 			}
 		}
 
 		// we skip lines that had a recognized command in them, or that
 		// aren't in a block
-		if isCommand || curBlockName == "" {
+		if isCommand || curBlock == nil {
 			continue
 		}
 
 		// add the line we processed to the current block we're in
 		// and account for having an existing curVal or not. If we
 		// don't then we assign curVal to start with the line we
 		// just found.
-		curVal, ok := blocks[curBlockName]
-		if ok {
-			blocks[curBlockName] = curVal + "\n" + line
+		if curBlock.Contents != "" {
+			curBlock.Contents += "\n" + line
 		} else {
-			blocks[curBlockName] = line
+			curBlock.Contents = line
+		}
+	}
+
+	if curBlock != nil {
+		return nil, fmt.Errorf("found dangling Block (%s) in %s", curBlock.Name, filePath)
+	}
+
+	if sourceTemplate != nil && sourceTemplate.adoptMode {
+		var err error
+		blocks, err = adoptBlocks(r, blocks, sourceTemplate)
+		if err != nil {
+			return nil, err
 		}
 	}
 
-	if curBlockName != "" {
-		return nil, fmt.Errorf("found dangling Block (%s) in %s", curBlockName, filePath)
+	return blocks, nil
+}
+
+// adoptBlocks adopts the blocks from the source template into the existing blocks
+func adoptBlocks(r io.ReadSeeker, blocks map[string]*blockInfo, sourceTemplate *Template) (map[string]*blockInfo, error) {
+	tr := bytes.NewReader(sourceTemplate.Contents)
+	templateBlocks, err := parseBlocksInner(tr, sourceTemplate.Path, nil)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
**thought**: I think this would technically not work for a template with generated blocks? I know Outreach used those for some cases and I'm not sure how I feel about that so I'm okay with making the decision to not support them for adopt mode, but we probably want to call that out if that's the case.

Reply:
Talked offline, this doesn't feel like an adopt use case worth the crazy complexity of trying to account for, at least for now.